### PR TITLE
no longer throw exception when xml is null (if we opened a json file)

### PR DESCRIPTION
**Pull request by mjkkirschner**
https://jira.autodesk.com/browse/QNTM-2122

### Purpose

This PR attempts to deal with a crash which occurs when opening any file which has unresolved nodes which do not contain XML. So in effect any JSON file which contains dummy nodes as of master branch. (that might change after the PR for 2080)...

This occurs because of an assertion that is made in the getter for `OriginalXMLcontent` if the content was not xml an exception was raised. 

This getter is invoked very frequently upon any node property change, undo, redo, analytics upload, node move, node click etc.

What happens now is that null is simply returned instead, I investigated converting the JSON to xml etc and this turned out to be a big pain, things got very complex fast.

If we simply return null though, copy paste and undo/redo will not work correctly, as the JSON content of the dummy node will be lost... I had to modify the deserialize and serialize methods of dummy node to save the JSON content as a string and then parse it back if it existed... I believe this is the simplest fix possible until we actually refactor undo/redo copy/paste to stop using XML... hopefully this is sooner rather than later. ⚠️ 😢 🔜 

I've added some tests that assert the JSON content exists after copy paste and undo... I think we need more tests, any suggestions for good ones?

I also noticed that analytics heart beat is converting the WS to XML quite frequently, I found this quite odd, and switched it to JSON conversion, note that this WS will be missing the view block, there is no way around that without moving this code to the DynamoViewModel, I didn't know enough about what it does to do that in this PR.

### Declarations

Check these if you believe they are true

- [ ] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@QilongTang 
@ColinDayOrg 
### FYIs

@ramramps ~~@saintentropy~~(sorry for the ping) @jnealb @smangarole

### CODE REVIEW

**Comment by mjkkirschner** [comment]
@ramramps note the builds are failing because of the libraryViewExtensions tests.

File: test/DynamoCoreTests/DummyNodeTests.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
+﻿using Dynamo.Graph.Nodes;
+using Newtonsoft.Json.Linq;
+using NUnit.Framework;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using static Dynamo.Models.DynamoModel;
```
Comment (ColinDayOrg) [inline test/DynamoCoreTests/DummyNodeTests.cs]: The static keyword here is causing the build checks to fail.

**Comment by ColinDayOrg** [comment]
Other than the failing CI checks, this LGTM.

File: src/DynamoCore/Graph/Nodes/DummyNode.cs [hunk 2]
```diff
@@ -227,6 +235,12 @@ private void SaveNode(XmlDocument xmlDoc, XmlElement nodeElement, SaveContext co
                     originalNode.AppendChild(nodeContent);
                     nodeElement.AppendChild(originalNode);
                 }
+                //if the node actually came from JSON we need to
+                //still preserve that incase this node was copied or undo/redone
+                else if (OriginalNodeContent is JObject)
+                {
+                    nodeElement.SetAttribute("OriginalNodeContent", OriginalNodeContent.ToString());
```
Comment (QilongTang) [inline src/DynamoCore/Graph/Nodes/DummyNode.cs]: I'm a bit confused here, so the XML save function will still be hit when we do undo/redo?

File: test/DynamoCoreTests/DummyNodeTests.cs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+            //assert same content after move.
+            Assert.AreEqual(contentPreMove, (dummyNode.OriginalNodeContent as JObject).ToString());
+
+            //undo the move.
+            CurrentDynamoModel.CurrentWorkspace.Undo();
+            Assert.AreEqual(originalPos, dummyNode.X);
+
+            //assert the content is the same as when we started.
+            Assert.AreEqual(contentPreMove, (dummyNode.OriginalNodeContent as JObject).ToString());
+        }
```
Comment (QilongTang) [inline test/DynamoCoreTests/DummyNodeTests.cs]: Do we have coverage on Dummy Node save and reopen content comparison in Json not talking about from XML?

**Comment by QilongTang** [comment]
This one looks much cleaner than the other one, although mixing XML serialization code with Json is bit a headache when understanding, so I am leaning towards not serializing the old content at all :thinking:

File: src/DynamoCore/Graph/Nodes/DummyNode.cs (hunk 2 above)
Comment (mjkkirschner) [reply src/DynamoCore/Graph/Nodes/DummyNode.cs]: yes

File: test/DynamoCoreTests/DummyNodeTests.cs (hunk 3 above)
Comment (mjkkirschner) [reply test/DynamoCoreTests/DummyNodeTests.cs]: I don't think so. I can add that.

**Comment by mjkkirschner** [comment]
@QilongTang if we don't serialize the old content at all here then copy/paste /undo/redo of dummy nodes will produce garbage when saved, I think we have to do something like this until we refactor copy/paste and undo/redo.

What do you guys think about the heartBeat serialization of the workspace to JSON

**Comment by ColinDayOrg** [comment]
I think that changing the heartbeat code to use JSON is a good idea, assuming that there is parity between the XML format and the JSON format.

**Comment by mjkkirschner** [comment]
@ColinDayOrg thanks for your thoughts, unfortunately in this case there is not parity, after talking with @QilongTang we might need to change the analytics end for that change, so I'll revert it for now and file a task.

**Comment by QilongTang** [comment]
@ColinDayOrg @mjkkirschner  https://jira.autodesk.com/browse/QNTM-2298

**Comment by mjkkirschner** [comment]
@QilongTang added another test - this PR is ready for another look.

**Comment by QilongTang** [comment]
LGTM

**Pull request merged**